File: sudoku-common/common-web/src/main/java/com/zumin/sudoku/common/web/handler/GlobalExceptionHandler.java
package com.zumin.sudoku.common.web.handler;

import static com.zumin.sudoku.common.core.enums.CommonStatusCode.ERROR;
import static com.zumin.sudoku.common.core.enums.CommonStatusCode.INVALID_REQUEST_PARAM_ERROR;

import cn.hutool.core.util.StrUtil;
import com.zumin.sudoku.common.core.exception.BaseException;
import com.zumin.sudoku.common.core.result.CommonResult;
import com.zumin.sudoku.common.web.exception.FormParameterConversionException;
import java.io.Serializable;
import java.util.stream.Collectors;
import javax.validation.ConstraintViolation;
import javax.validation.ConstraintViolationException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.autoconfigure.condition.ConditionalOnExpression;
import org.springframework.context.support.DefaultMessageSourceResolvable;
import org.springframework.validation.BindException;
import org.springframework.validation.ObjectError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

/**
 * 全局异常处理器
 */
@Slf4j
@RestControllerAdvice
@ConditionalOnExpression("#{!'false'.equals(environment['common.web.global.exception-handler'])}")
public class GlobalExceptionHandler {

  private static final String INVALID_PARAM_TIP_TEMPLATE = "{}：{}";

  /**
   * 处理所有异常
   *
   * @param e 异常
   * @return 经过包装的结果对象
   */
  @ExceptionHandler(value = Exception.class)
  public CommonResult<Exception> exceptionHandler(Exception e) {
    log.error("[Exception]", e);
    return CommonResult.error(ERROR);
  }

  /**
   * 处理基础异常
   *
   * @param e 异常
   * @return 经过包装的结果对象
   */
  @ExceptionHandler(value = BaseException.class)
  public CommonResult<BaseException> baseExceptionHandler(BaseException e) {
    log.debug("[" + e.getClass().getSimpleName() + "]", e);
    return CommonResult.error(e.getStatusCode());
  }

  /**
   * 处理表单参数转换异常
   *
   * @param e 登录异常
   * @return 经过包装的结果对象
   */
  @ExceptionHandler(value = FormParameterConversionException.class)
  public CommonResult<FormParameterConversionException> formParameterConversionExceptionHandler(FormParameterConversionException e) {
    log.warn("[FormParameterConversionException]", e);
    return CommonResult.error(e.getStatusCode());
  }

  /**
   * 处理参数校验异常
   *
   * @param e 参数校验异常
   * @return 经过包装的结果对象
   */
  @ExceptionHandler(value = ConstraintViolationException.class)
  public CommonResult<ConstraintViolationException> constraintViolationExceptionHandler(ConstraintViolationException e) {
    log.debug("[FormParameterConversionException]", e);
    return buildInvalidParamErrorCommonResult(
        e.getConstraintViolations().stream().map(ConstraintViolation::getMessage).collect(Collectors.joining(";")));
  }

  /**
   * 处理参数校验异常
   *
   * @param e 参数校验异常
   * @return 经过包装的结果对象
   */
  @ExceptionHandler(value = BindException.class)
  public CommonResult<java.net.BindException> bindExceptionHandler(BindException e) {
    log.debug("[ParameterVerificationException]", e);
    return buildInvalidParamErrorCommonResult(
        e.getAllErrors().stream().map(ObjectError::getDefaultMessage).collect(Collectors.joining(";")));
  }

  /**
   * 处理参数校验异常
   *
   * @param e 参数校验异常
   * @return 经过包装的结果对象
   */
  @ExceptionHandler(value = MethodArgumentNotValidException.class)
  public CommonResult<MethodArgumentNotValidException> bindExceptionHandler(MethodArgumentNotValidException e) {
    log.debug("[ParameterVerificationException]", e);
    return buildInvalidParamErrorCommonResult(
        e.getBindingResult().getAllErrors().stream().map(DefaultMessageSourceResolvable::getDefaultMessage).collect(
            Collectors.joining(";")));
  }

  /**
   * 构建无效的参数的返回结果
   *
   * @param error 异常原因
   * @param <T>   异常类型
   * @return 经过包装的结果对象
   */
  private <T extends Serializable> CommonResult<T> buildInvalidParamErrorCommonResult(String error) {
    String tip = null;
    if (StrUtil.isNotBlank(error)) {
      tip = StrUtil.format(INVALID_PARAM_TIP_TEMPLATE, INVALID_REQUEST_PARAM_ERROR.getMessage(), error);
    }
    return CommonResult.error(INVALID_REQUEST_PARAM_ERROR, tip);
  }
}

File: sudoku-common/common-web/src/main/java/com/zumin/sudoku/common/web/handler/GlobalExceptionHandler.kt
package com.zumin.sudoku.common.web.handler

import com.alibaba.csp.sentinel.slots.block.BlockException
import com.zumin.sudoku.common.core.CommonResult
import com.zumin.sudoku.common.core.CommonResult.Companion.error
import com.zumin.sudoku.common.core.code.CommonStatusCode
import com.zumin.sudoku.common.core.exception.BaseException
import com.zumin.sudoku.common.web.exception.FormParameterConversionException
import mu.KotlinLogging
import org.springframework.boot.autoconfigure.condition.ConditionalOnExpression
import org.springframework.validation.BindException
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import java.io.Serializable
import javax.validation.ConstraintViolationException

/**
 * 全局异常处理器
 */
@RestControllerAdvice
@ConditionalOnExpression("#{!'false'.equals(environment['common.web.exception-handler'])}")
class GlobalExceptionHandler {
  /**
   * 处理所有异常
   *
   * @param e 异常
   * @return 经过包装的结果对象
   */
  @ExceptionHandler(value = [Exception::class])
  fun exceptionHandler(e: Exception): CommonResult<Exception> {
    logger.error("[Exception]", e)
    return error(CommonStatusCode.ERROR)
  }

  /**
   * 处理基础异常
   *
   * @param e 异常
   * @return 经过包装的结果对象
   */
  @ExceptionHandler(value = [BaseException::class])
  fun baseExceptionHandler(e: BaseException): CommonResult<BaseException> {
    logger.debug("[${e.javaClass.simpleName}]", e)
    return error(e.statusCode)
  }

  /**
   * 处理流量控制异常
   *
   * @param e 异常
   * @return 经过包装的结果对象
   */
  @ExceptionHandler(value = [BlockException::class])
  fun baseExceptionHandler(e: BlockException): CommonResult<BlockException> {
    logger.debug("[${e.javaClass.simpleName}]", e)
    return error(CommonStatusCode.ACCESS_DENIED)
  }

  /**
   * 处理表单参数转换异常
   *
   * @param e 登录异常
   * @return 经过包装的结果对象
   */
  @ExceptionHandler(value = [FormParameterConversionException::class])
  fun formParameterConversionExceptionHandler(e: FormParameterConversionException): CommonResult<FormParameterConversionException> {
    logger.warn("[FormParameterConversionException]", e)
    return error(e.statusCode)
  }

  /**
   * 处理参数校验异常
   *
   * @param e 参数校验异常
   * @return 经过包装的结果对象
   */
  @ExceptionHandler(value = [ConstraintViolationException::class])
  fun constraintViolationExceptionHandler(e: ConstraintViolationException): CommonResult<ConstraintViolationException> {
    logger.debug("[FormParameterConversionException]", e)
    return buildInvalidParamErrorCommonResult(e.constraintViolations.map { it.message }.joinToString(";"))
  }

  /**
   * 处理参数校验异常
   *
   * @param e 参数校验异常
   * @return 经过包装的结果对象
   */
  @ExceptionHandler(value = [BindException::class])
  fun bindExceptionHandler(e: BindException): CommonResult<java.net.BindException> {
    logger.debug("[ParameterVerificationException]", e)
    return buildInvalidParamErrorCommonResult(e.allErrors.map { it.defaultMessage }.joinToString(";"))
  }

  /**
   * 处理参数校验异常
   *
   * @param e 参数校验异常
   * @return 经过包装的结果对象
   */
  @ExceptionHandler(value = [MethodArgumentNotValidException::class])
  fun bindExceptionHandler(e: MethodArgumentNotValidException): CommonResult<MethodArgumentNotValidException> {
    logger.debug("[ParameterVerificationException]", e)
    return buildInvalidParamErrorCommonResult(e.bindingResult.allErrors.map { it.defaultMessage }.joinToString(";"))
  }

  /**
   * 构建无效的参数的返回结果
   *
   * @param errorStr 异常原因
   * @param <T>   异常类型
   * @return 经过包装的结果对象
   */
  private fun <T : Serializable?> buildInvalidParamErrorCommonResult(errorStr: String): CommonResult<T> {
    var tip: String? = null
    if (errorStr.isNotBlank()) {
      tip = "${CommonStatusCode.INVALID_REQUEST_PARAM_ERROR.getMessage()}：${errorStr}"
    }
    return error(CommonStatusCode.INVALID_REQUEST_PARAM_ERROR, tip)
  }
}

private val logger = KotlinLogging.logger { }